[MRG] Fixed assigning of empty values to data elements

mrbean-bremen opened the pull request:
- setting a value to None or an empty string now always results
  in an empty tag (VM = 0); see #896

This was a bit more tricky than I thought, and I'm not sure if this is the best way to implement this.
Also I'm not sure if this may cause some compatibility issues...
Anyway, I'm posting this as a discussion base, maybe someone has some better ideas.

Ping @scaramallion, @darcymason

pep8speaks commented:
Hello @mrbean-bremen! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:











* In the file [`pydicom/tests/test_multival.py`](https://github.com/pydicom/pydicom/blob/77a0df9eb679135f63d2f035f0f1b80ecc2d6fba/pydicom/tests/test_multival.py):

> [Line 139:54](https://github.com/pydicom/pydicom/blob/77a0df9eb679135f63d2f035f0f1b80ecc2d6fba/pydicom/tests/test_multival.py#L139): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 139:55](https://github.com/pydicom/pydicom/blob/77a0df9eb679135f63d2f035f0f1b80ecc2d6fba/pydicom/tests/test_multival.py#L139): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

mrbean-bremen commented on pydicom/multival.py:
```diff
@@ -2,13 +2,17 @@
     from collections import MutableSequence
 
 
+EMPTY_VAL = ''
```
This is a leftover I forgot to remove. I have tried to print some special value for empty tags, but that  broke a lot of code.

mrbean-bremen commented on pydicom/tests/test_dataelem.py:
```diff
@@ -368,6 +369,62 @@ def test_private_repeater_tag(self):
+            elem = ds[tag_for_keyword(kw)]
+            assert bool(elem.value) is False
+            assert elem.VM == 0
+
```
Took that directly from the issue :)

scaramallion commented on pydicom/dataelem.py:
```diff
@@ -166,7 +172,14 @@ def __init__(self,
-        if already_converted:
+        if value is None:
+            if VR in INT_VR_VALUES:
+                self._value = MultiValue(int, [])
```
I'm not sure if this is the best way to do this since it fails the reflexive test

```python
ds.SomeElement = None
assert ds.SomeElement == None
```

darcymason replied:
I agree with the reflexive test here, anything set to None should ideally be kept as None, if that is possible.

mrbean-bremen replied:
I partly agree. `None` is a natural value representation for empty type 2 elements, so using it instead of empty arrays makes sense. For numerical elements this should probably be the returned value. For text values, I'm not so sure, as they can also be represented by an empty string - there is no difference between an empty value and no value, so we have to decide what to use, and I tend to use an empty string as being less disruptive. It is a possibility to adapt comparisons to return True both for empty strings and `None`, though I'm not sure about that either.
Also keep in mind that after saving and reloading a dataset we cannot know if we originally assigned an empty string, an empty array or `None` to a text element, as they are all equivalent - we have to decide for a default representation. I may have to think more about this...

mrbean-bremen replied:
I partly agree. `None` is a natural value representation for empty type 2 elements, so using it instead of empty arrays makes sense. For numerical elements this should probably be the returned value. For text values, I'm not so sure, as they can also be represented by an empty string - there is no difference between an empty value and no value, so we have to decide what to use, and I tend to use an empty string as being less disruptive. It is a possibility to adapt comparisons to return True both for empty strings and `None`, though I'm not sure about that either.
Also keep in mind that after saving and reloading a dataset we cannot know if we originally assigned an empty string, an empty array or `None` to a text element, as they are all equivalent - we have to decide for a default representation. I may have to think more about this...

mrbean-bremen commented:
Note - I rebased with master and squashed it all together to get rid of all the merge conflicts and duplicate commits.

mrbean-bremen commented:
I'm setting this to MRG for another review and to discuss what has to be done in the scope of this PR, and what can possibly be done later.

mrbean-bremen commented:
Ok, I changed the behavior:
- the value is retained after setting it to `None`, `''` or `[]`
- an excpeption is setting a text value to `b''` - this is handled for some VRs by decoding it to a `str`, and I didn't change that
- I added a single config value for the empty value set after decoding, currently set to `None`, e.g. all empty values are decoded as `None`

@darcymason - I think this is closer to what you had in mind, though I'm not sure if this sufficient.
Please have a look!

scaramallion commented on pydicom/config.py:
```diff
@@ -60,6 +60,10 @@ def DS_decimal(use_Decimal_boolean=True):
 """
 
+empty_value = None
+"""Defines the value that an empty data element is set to
```
Maybe we should start by defaulting to `''` to maintain current behaviour and put a comment in the release notes that the default behaviour will change in v1.5 or v1.6.

mrbean-bremen replied:
Yes, I've thought about this, too. That would certainly be less disrupting.

mrbean-bremen replied:
Ok, done.

mrbean-bremen replied:
I have to re-think how to configure this, as besically each possible configuration has to be tested. So, maybe I will just support setting `''` or `None` to text VR values, and always set `None` to the rest. Maybe allow `b''` for binary values optionally, too.

scaramallion commented:
Can you add a simple test to confirm that changing `config.empty_value` gives the expected result?

mrbean-bremen commented:
> Can you add a simple test to confirm that changing config.empty_value gives the expected result?

Sure, I wanted to do this, but obviously forgot it.

mrbean-bremen commented:
Ok, now only two options are available: decode all empty values as `None`, or decode text values as `''` and other values as `None` (set as default as suggested by @scaramallion). Setting a value manually still retains that value.

darcymason commented on pydicom/config.py:
```diff
@@ -86,9 +86,12 @@ def DS_decimal(use_Decimal_boolean=True):
+If ``False`` (the default), the value of an empty data element with 
+a text VR is an empty string, for all other VRs it is also ``None``.
+Note that the default of this value will change to ``True`` in version 1.5.
+"""
```
Perhaps version 2.0 based on discussion in [Python 2 support thread](https://github.com/pydicom/pydicom/issues/872#issuecomment-516194302)?

mrbean-bremen replied:
Yes, I wasn't sure about this anyway. 2.0 sounds good.

darcymason commented on pydicom/dataelem.py:
```diff
@@ -42,6 +42,42 @@
+        return None
+    if VR in ('AE', 'AS', 'CS', 'DA', 'DT', 'LO', 'LT',
+              'PN', 'SH', 'ST', 'TM', 'UC', 'UI', 'UR', 'UT'):
+        return b'' if raw else ''
```
I went to see if there is a list like this already defined, and in valuerep.py there is a list called `text_VRs` but much shorter -- it is only those to which specific character set apply.  I don't have an answer, but it would be nice to have more consistency in what we mean by 'text VR'.

mrbean-bremen replied:
Ah well, I'm not good with names... I have been thinking about a bunch of constants with VR tuples, such as `numeric_VRs` with subsets `binary_numeric_VRs` and `textual_numeric_VRs` and so on, but we actually don't need them. 
The `text_VRs` tuple is there because I had introduced `convert_text` for these additionally to the existing `convert_string` method (which didn't handle encodings correctly), but I agree it is a bit unclear.
Right now I have no good idea, I'll think about it...

mrbean-bremen commented on pydicom/config.py:
```diff
@@ -86,9 +86,12 @@ def DS_decimal(use_Decimal_boolean=True):
-"""Defines the value that an empty data element is set to
-after decoding."""
+use_none_as_empty_value = False
+""" If ``True``, the value of decoded empty data element is always ``None``.
```
Agreed, will change this.

mrbean-bremen commented on pydicom/dataelem.py:
```diff
@@ -323,19 +350,37 @@ def value(self, val):
+
+    def clear(self):
+        """Clears the value, e.g. sets it to `None`."""
+        self._value = None
```
Actually, I forgot to change this - this has also to be set to the configured empty value.

scaramallion commented:
Looks good
